helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -163,25 +157,39 @@ protected static RealmAwareZkClient createZkClient(HelixManager manager) {
       }
     }
 
-    return SharedZkClientFactory.getInstance().buildZkClient(
-        new HelixZkClient.ZkConnectionConfig(zkAddress),
-        clientConfig.createHelixZkClientConfig().setZkSerializer(new ZNRecordSerializer()));
+    // Fail early instead of retrying and blocking TaskStateModelFactory creation
+    HelixZkClient.ZkClientConfig zkClientConfig =
+        clientConfig.createHelixZkClientConfig().setZkSerializer(new ZNRecordSerializer());
+    zkClientConfig.setOperationRetryTimeout((long) 60000);
+    return SharedZkClientFactory.getInstance()
+        .buildZkClient(new HelixZkClient.ZkConnectionConfig(zkAddress), zkClientConfig);
```

Comment:
This might be problematic (only applies to SharedZkClient):

```
  /**
   * Construct and init a ZkConnection Manager.
   *
   * @param zkConnection
   * @param connectionTimeout
   */
  protected ZkConnectionManager(IZkConnection zkConnection, long connectionTimeout,
      String monitorKey) {
    super(zkConnection, (int) connectionTimeout, HelixZkClient.DEFAULT_OPERATION_TIMEOUT,
        new BasicZkSerializer(new SerializableSerializer()), MONITOR_TYPE, monitorKey, null, true);
    _monitorKey = monitorKey;
    LOG.info("ZkConnection {} was created for sharing.", _monitorKey);
  }
```

Even if we set and pass in a zkClientConfig, it just gets ignored and will take on a default value, which is infinite. This is a limitation for shared zk client - I'm okay with not addressing this in this PR, but we should document this erroneous behavior.

Also, is there a reason we're not giving a client config with timeout to FederatedZkClient?

Reply:
Strange. I will revert these lines then, since setting the retry timeout is obsolete. Issue created: https://github.com/apache/helix/issues/1682.

I'm not adding a timeout to FederatedZkClient because it creates new connections, instead of possibly reuses old connections like SharedZkClient. I don't see the possibility of newly created connections having mismatching zxid problems and hang infinitely; if that's a concern, then `ZkBaseDataAccessor` or `ZkBucketDataAccessor` should also set a retry timeout. In short, since it creates new connections, I don't see it as necessary to go beyond the default behavior and set retry timeouts.

Reply:
My suggestion is to give it a valid zk client config with a non-infinite operation retry timeout regardless of what type of ZkClient we're using here.

Operation retry timeout and the server's refusal due to lagging zxid seen are different modes of failure. Other Helix APIs like you mentioned can accept client config at creation - those are managed by users.

Reply:
Discussed offline: there are values in preventing infinite hanging on client side. Added retry timeout to both.